### PR TITLE
Simple planar-renderer script to render planar reflections

### PR DESCRIPTION
**Pull request by mvaligursky**
- PlanarRenderer script, which handles mirroring by plane and rendering to texture
- made Plane class public
- renamed couple of examples to REFLECTION*** pattern to keep those grouped.

TODO:
- integrate the result to standard material, allowing normal maps, roughness ... (separate PR)

Screenshot 2023-03-23 at 15 44 37

### CODE REVIEW

File: src/core/shape/plane.js [hunk 1]
```diff
@@ -39,6 +37,16 @@ class Plane {
         }
     }
 
+    /**
+     * The d term for a scalar equation of the form ax + by + cz + d = 0, where a, b and c
+     * are the x, y and z components of the plane's normal.
+     *
+     * @type {number}
+     */
+    get d() {
```
Comment (slimbuck) [inline src/core/shape/plane.js:46]: shouldn't we fix plane before making it public api? why on earth it stores a normal and point is beyond me!

Comment (mvaligursky) [reply src/core/shape/plane.js]: yep, that makes sense and done. Larger PR, but hey.

File: src/framework/components/element/element-drag-helper.js [hunk 2]
```diff
@@ -158,22 +160,13 @@ class ElementDragHelper extends EventHandler {
...
-
-            _entityRotation.copy(this._element.entity.getRotation()).invert().transformVector(position, position);
-
-            position.mul(this._dragScale);
-
-            return position;
+        if (_plane.intersectsRay(_ray, _point)) {
+            _entityRotation.copy(this._element.entity.getRotation()).invert().transformVector(_point, _point);
+            _point.mul(this._dragScale);
+            return _point;
```
Comment (mvaligursky) [inline src/framework/components/element/element-drag-helper.js:169]: I tested this change in the engine example: scroll-view

File: src/core/shape/plane.js [hunk 3]
```diff
@@ -109,7 +107,8 @@ class Plane {
      * @returns {this} A duplicate Plane.
      */
     clone() {
-        return new this.constructor(this.point, this.normal);
+        const clone = new this.constructor();
+        return clone.copy(this);
```
Comment (willeastcott) [inline src/core/shape/plane.js:111]:
```suggestion
        /** @type {this} */
        const cstr = this.constructor;
        return new cstr().copy(this);
```

Comment (willeastcott) [reply src/core/shape/plane.js:111]: This seems to be the format for `clone` across the engine.

Comment (mvaligursky) [reply src/core/shape/plane.js:111]: hah, I copied from DepthState ;)

**Comment by GSterbrant** [comment]
Why not make the example default to using a 100% scaled reflection buffer? Looks less aliased that way 🙂.

**Review by GSterbrant — APPROVED** [review]

**Comment by mvaligursky** [comment]
> Why not make the example default to using a 100% scaled reflection buffer? Looks less aliased that way 🙂.

Sure, done.

File: src/core/math/mat4.js [hunk 4]
```diff
@@ -778,6 +778,42 @@ class Mat4 {
...
     }
 
+    /**
+     * Sets the matrix to a reflection matrix, which can be used as a mirror transformation by the
+     * plane.
+     *
+     * @param {import('./plane.js').Plane} plane - The plane to reflect by.
+     * @returns {Mat4} Self for chaining.
+     */
+    setReflection(plane) {
```
Comment (willeastcott) [inline src/core/math/mat4.js:788]: I'm a bit late to the party here, but this is introducing a dependency of the math API on the shape API. I'm not a big fan of that.

File: src/core/shape/plane.js [hunk 5]
```diff
@@ -1,54 +1,48 @@
...
+     * Defaults to {@link pc.Vec3#ZERO}.
      */
-    constructor(point, normal) {
-        if (point) {
-            this.point.copy(point);
-        }
-        if (normal) {
-            this.normal.copy(normal);
-        }
+    constructor(normal = Vec3.UP, point = Vec3.ZERO) {
```
Comment (willeastcott) [inline src/core/shape/plane.js:32]: Should we have changed the constructor to take a normal and d?

File: src/core/math/mat4.js (hunk 4 above)
Comment (mvaligursky) [reply src/core/math/mat4.js:788]: Good point.
I feel like moving Plane to Math. I even assumed it's there.

File: src/core/shape/plane.js [hunk 6]
```diff
@@ -1,54 +1,48 @@
...
+     * Sets the plane to use a specified normal and pass through a specified point.
      *
-     * @param {Vec3} point - The starting point of the plane.
      * @param {Vec3} normal - The normal of the plane.
+     * @param {Vec3} point - The point on a plane.
      * @returns {Plane} Self for chaining.
      */
-    set(point, normal) {
-        this.point.copy(point);
+    set(normal, point) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/core/shape/plane.js:43]: yep, same conclusion here. On it.

**Comment by willeastcott** [comment]
Wondering whether `Frustum` should be using this new `Plane` class...

File: src/core/shape/plane.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/core/shape/plane.js:43]: done here https://github.com/playcanvas/engine/pull/5196

**Pull request merged**